src/intrinsics/mod.rs

```diff
@@ -372,8 +399,28 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                     _ => bug!(),
                 };
                 let res = this.binary_op(op, &a, &b)?;
-                // `binary_op` already called `generate_nan` if necessary.
-                this.write_immediate(*res, dest)?;
+                // `binary_op` already called `generate_nan` if needed
+                // Apply a relative error of 16ULP to simulate non-determinism
+                fn apply_error_and_write<'tcx, F: Float + Into<Scalar> >(
+                    ecx: &mut MiriInterpCx<'tcx>,
+                    res: F,
+                    dest: &MPlaceTy<'tcx>
+                ) -> InterpResult<'tcx> {
```

Comment:
Please add a helper function that does `apply_random_float_error_ulp` on an `ImmTy`, so that the `match` can be shared between here and the `*_fast` intrinsics below. Keep the `write_scalar` separate.